websearch/tavily/src/lib.rs

```diff
@@ -0,0 +1,160 @@
+mod client;
+mod conversions;
+
+use std::cell::RefCell;
+
+use crate::client::{SearchRequest, TavilySearchApi};
+use crate::conversions::{params_to_request, response_to_results, validate_search_params};
+use golem_web_search::golem::web_search::web_search::{
+    Guest, GuestSearchSession, SearchError, SearchMetadata, SearchParams, SearchResult,
+    SearchSession,
+};
+
+use golem_web_search::LOGGING_STATE;
+
+struct TavilySearch {
+    client: TavilySearchApi,
+    request: SearchRequest,
+    params: SearchParams,
+    finished: bool,
+    metadata: Option<SearchMetadata>,
+    current_page: u32,
+}
+
+impl TavilySearch {
+    fn new(client: TavilySearchApi, request: SearchRequest, params: SearchParams) -> Self {
+        Self {
+            client,
+            request,
+            params,
+            finished: false,
+            metadata: None,
+            current_page: 1,
+        }
+    }
+
+    fn next_page(&mut self) -> Result<Vec<SearchResult>, SearchError> {
+        if self.finished {
+            return Ok(vec![]);
+        }
+
+        let request = self.request.clone();
+        // Note: Tavily's SearchRequest doesn't have pagination fields
+        // We'll use the existing request and track pagination in metadata
+
+        let response = self.client.search(request)?;
+        let (results, metadata) = response_to_results(response, &self.params, self.current_page);
+
+        // Check if more results are available
+        if let Some(ref meta) = metadata {
+            // Check if we got the full count requested
+            let max_results = self.params.max_results.unwrap_or(10);
+            let has_more_results = results.len() == (max_results as usize);
+
+            // Also check if next_page_token is available
+            let has_next_page = meta.next_page_token.is_some();
+
+            // Only set finished if no more results available
+            self.finished = !has_more_results || !has_next_page;
+
+            // Increment page for next request if not finished
+            if !self.finished {
+                self.current_page += 1;
+            }
+        } else {
+            self.finished = true;
```

Comment:
If we know that there are no more pages, we can skip these checks and just always set self.finished = true